### PR TITLE
Allow kube-proxy could return unhealthy if endpoints/services don't update

### LINKED ISSUE #76338: Kube-proxy hangs on api-server errors
**What happened**:
After some api-server outages (2 apparent in logs and they lastestd minutes at least)
Kube proxy stopped recieving updates from endpoionts. And stopped updating the ip tables
Here's the kube-proxy logs which show errors and then nothing for 4 days (Log ended on 4/8/2019)
[kubehunglogs.txt](https://github.com/kubernetes/kubernetes/files/3060250/kubehunglogs.txt)
Here's the iptables and endpoints that were out of wack.  
[iptablesandendpoints.txt](https://github.com/kubernetes/kubernetes/files/3060257/iptablesandendpoints.txt)

We've had this happen twice so reporting it now though we're still working on a repo. 
The core dns service being out of date is what really kills us as the dns lookup will just start hanging. 

**What you expected to happen**:
Either kube-proxy should continue retrying or fail so it can be restarted. 

**How to reproduce it (as minimally and precisely as possible)**:
Take an api-server down for several minutes? We may try this today. Not oure api server is an aks instance where it is acually fqdn not a service in the cluster. 

**Anything else we need to know?**:
Sadly the pod was restarted but we're curious if the healthz endpoint may help here. Maybe one stargety is check the last update time of the  if the lastUpdated  time from healthz and see if we can  make a liveness prove kill kube-proxy if it has been hours since last update.                       

**Environment**:
- Kubernetes version (use `kubectl version`): 1.12.5
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): Ubuntu 16.04.5 LTS
- Kernel (e.g. `uname -a`):4.15.0-1041-azure

### PR DESCRIPTION
**Pull request by paulgmiller**
We have an issue in our cluster where watches can hang after getting disconnected from the api server (we're guessing there is some comonent in between that doesn't close the connection when the api server cycles). This would let kube-proxy fix itself rather than just slowly rotting in a hard to discover way. 

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
 /kind bug

**What this PR does / why we need it**:
If watch hangs kube proxy can get quite out of date. See
https://github.com/kubernetes/kubernetes/issues/76338
While watch not hanging is obviously good I was wondering if it might makes sense to actually get this info into the health check so a liveness probe can action on it. 

**Which issue(s) this PR fixes**:

Fixes #76338

**Special notes for your reviewer**:
To this is just a proof on concept if we decided we did want this in the health check we could add.
1) a wrapper proxyconfig.EndpointsHandler that updated on api calls rather than having each proxier do it.
2) allow the livness probe to take timeout as a header so you could onlt make it fail if you wanted to be sensitive to api updates. 

**Does this PR introduce a user-facing change?**:

```release-note
endpointsUpdateTimeout and servicesUpdateTimeout headers can be passed to kube-proxy healthz and will fail the probe if endpoints or services haven't been updated in the duration in the header
```

### CODE REVIEW

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -538,6 +539,9 @@ func (proxier *Proxier) OnEndpointsAdd(endpoints *v1.Endpoints) {
 
 func (proxier *Proxier) OnEndpointsUpdate(oldEndpoints, endpoints *v1.Endpoints) {
 	if proxier.endpointsChanges.Update(oldEndpoints, endpoints) && proxier.isInitialized() {
+		if proxier.healthzServer != nil {
```
Comment (paulgmiller) [inline pkg/proxy/iptables/proxier.go]: could wrap proxyconfig.EndpointsHandler so each proxier didn't need this if we decided it was useful.

**Comment by paulgmiller** [comment]
@thockin who made the last big change on health checks to see if he has an opinion.

File: pkg/proxy/healthcheck/healthcheck.go [hunk 2]
```diff
@@ -335,13 +342,20 @@ func (h healthzHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
 	if val := h.hs.lastUpdated.Load(); val != nil {
 		lastUpdated = val.(time.Time)
 	}
+	apiLastUpdated := time.Time{}
+	if val := h.hs.apiLastUpdated.Load(); val != nil {
+		apiLastUpdated = val.(time.Time)
```
Comment (JieJhih) [inline pkg/proxy/healthcheck/healthcheck.go]: It would be better to handle type assertion error.

Comment (paulgmiller) [reply pkg/proxy/healthcheck/healthcheck.go]: Ack think I can do that better

**Comment by paulgmiller** [comment]
Was going to tackle metrics but seems like Casey Callendrello did that back in february but it's not in 1.14 or 1.13 yet. 

commit 017f57a6b0cc052cfb041bf653f1ba984c6ec26b
Author: Casey Callendrello <cdc@redhat.com>
Date:   Wed Feb 13 18:48:45 2019 +0100

    proxy: add some useful metrics

    This adds some useful metrics around pending changes and last successful
    sync time.

    The goal is for administrators to be able to alert on proxies that, for
    whatever reason, are quite stale.

    Signed-off-by: Casey Callendrello <cdc@redhat.com>

**Comment by paulgmiller** [comment]
@dcbw who changed the endpoint/service call back code recently (4b07f80d204) to see if he's got an opinion on wether this useful.

**Comment by dcbw** [comment]
@paulgmiller can you add a release note label as well?

**Pull request closed without merging**